book_service/views.py

```diff
@@ -1,9 +1,78 @@
-from rest_framework import viewsets
+from drf_spectacular.utils import extend_schema, OpenApiParameter
+from rest_framework import viewsets, status
+from rest_framework.decorators import action
+from rest_framework.permissions import IsAdminUser
+from rest_framework.response import Response
 
 from book_service.models import Book
-from book_service.serializers import BookSerializer
+from book_service.permissions import IsAdminOrReadOnly
+from book_service.serializers import (
+    BookSerializer,
+    BookImageSerializer,
+    BookListSerializer,
+    BookDetailSerializer,
+)
 
 
 class BookViewSet(viewsets.ModelViewSet):
+    """Endpoints of the books in library with basic CRUD operations"""
+
     queryset = Book.objects.all()
     serializer_class = BookSerializer
+    permission_classes = [IsAdminOrReadOnly]
+    search_fields = ["title", "author"]
+    filterset_fields = ["cover"]
+
+    def get_serializer_class(self):
+        if self.action == "list":
+            return BookListSerializer
+
+        if self.action == "retrieve":
+            return BookDetailSerializer
+
+        if self.action == "upload_image":
+            return BookImageSerializer
+
+        return BookSerializer
```

Comment:
Here you can return self.serializer_class, it would look more pythonic

Reply:
Yep if you already have self.serializer then why not to use it )